core/mutex.c

```diff
@@ -32,42 +33,38 @@
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
-int _mutex_lock(mutex_t *mutex, volatile uint8_t *blocking)
+int mutex_lock(mutex_t *mutex)
 {
-    unsigned irqstate = irq_disable();
+    unsigned irq_state = irq_disable();
 
     DEBUG("PID[%" PRIkernel_pid "]: Mutex in use.\n", thread_getpid());
 
     if (mutex->queue.next == NULL) {
         /* mutex is unlocked. */
         mutex->queue.next = MUTEX_LOCKED;
-        DEBUG("PID[%" PRIkernel_pid "]: mutex_wait early out.\n",
+        DEBUG("PID[%" PRIkernel_pid "]: mutex_wait_and_lock early out.\n",
               thread_getpid());
-        irq_restore(irqstate);
-        return 1;
+        irq_restore(irq_state);
+        return 0;
     }
-    else if (*blocking) {
-        thread_t *me = thread_get_active();
-        DEBUG("PID[%" PRIkernel_pid "]: Adding node to mutex queue: prio: %"
-              PRIu32 "\n", thread_getpid(), (uint32_t)me->priority);
-        sched_set_status(me, STATUS_MUTEX_BLOCKED);
-        if (mutex->queue.next == MUTEX_LOCKED) {
-            mutex->queue.next = (list_node_t *)&me->rq_entry;
-            mutex->queue.next->next = NULL;
-        }
-        else {
-            thread_add_to_list(&mutex->queue, me);
-        }
-        irq_restore(irqstate);
-        thread_yield_higher();
-        /* We were woken up by scheduler. Waker removed us from queue.
-         * We have the mutex now. */
-        return 1;
+
+    thread_t *me = thread_get_active();
+    DEBUG("PID[%" PRIkernel_pid "]: Adding node to mutex queue: prio: %"
+          PRIu32 "\n", thread_getpid(), (uint32_t)me->priority);
+    sched_set_status(me, STATUS_MUTEX_BLOCKED);
+    if (mutex->queue.next == MUTEX_LOCKED) {
```

Comment:
So `MUTEX_LOCKED` marks the end of the list

Reply:
No, the actual semantics is that the mutex is a list of waiters for the mutex - unless there are no waiters. In that case it is either unlocked or locked.

Or graphically:

```
Unlocked mutex:
+-------+
| Mutex | --> NULL
+-------+

Locked mutex, no waiters:
+-------+
| Mutex | --> MUTEX_LOCKED
+-------+

Locked mutex, one waiter:
+-------+     +--------+
| Mutex | --> | Waiter | --> NULL
+-------+     +--------+

Locked mutex, 2 waiters:
+-------+     +--------+     +--------+
| Mutex | --> | Waiter | --> | Waiter | --> NULL
+-------+     +--------+     +--------+
```

It would make sense to document this :-) Let me do so later todo.

(The semantics of this cannot be changed for now, as the xtimer's own implementation of mutex_lock has opinions on this. But once all the clean up PRs are merged, no code outside of `core/mutex` will directly touch the internals and the encoding could be changed. However, there is a good reason for this special semantics: The list functions in `core` expect the list to be `NULL` terminated.)

Reply:
ping :wink:

Reply:
[pong](https://github.com/RIOT-OS/RIOT/pull/15398)!